Proposal for separation of ingest and database concerns in LinkML model

markquintontulloch opened the pull request:
As a result of a discussion within the A-Team, we believe the best way forward or the LinkML model is to separate ingest modelling from database / Java code modelling.  This is an initial proposed remodel for the classes currently ingested by the A-Team's curation database.  This accurately represents both our ingest DTO classes and our entity classes (used to generate the database schema via Hibernate).  It allows for fields to be required in the database but not in the ingest files, and allows both forward- and back-references to be explicitly described in the LinkML model.  

@oblodgett, @chris-grove  - Any initial feedback before taking this to the wider modelling group?

sierra-moxon commented:
I think this is excellent!!

sierra-moxon commented on model/schema/core.yaml:
```diff
@@ -156,28 +202,40 @@ classes:
+      note_type_name:
+        required_true
+
   SlotAnnotation:
```
remove this class?

markquintontulloch replied:
Yeah, I think there's probably a bunch of cleanup to be done before moving this PR from draft - just wanted to get feedback first

markquintontulloch commented:
@sierra-moxon @chris-grove The proposal has now been updated to use SlotAnnotations again (with one less stage of inheritance).  For the slot names representing the 'backlinks' I have kept the names the same rather than adding '_slot_annotation', e.g. 'allele_mutation_types' vs. 'allele_mutation_type_slot_annotations' - changing them seemed excessively verbose but maybe this would be preferred for the sake of clarity (I don't feel particularly strongly either way).  I did change the SlotAnnotation class names though just to make the design pattern a bit more obvious without relying on noticing the `is_a: SlotAnnotation`.

chris-grove commented on model/schema/affectedGenomicModel.yaml:
```diff
@@ -47,23 +47,52 @@ classes:
+      - subtype_name
+      - reference_curies
+      - data_provider
+      - reference_curies
```
slot 'reference_curies' present twice

chris-grove commented on model/schema/allele.yaml:
```diff
@@ -54,6 +54,23 @@ classes:
+      - allele_construct_associations
+      - allele_cell_line_associations
+      - allele_image_associations
+      - allele_origin_associations
```
missing 'allele_generation_method_associations' slot

chris-grove commented on model/schema/core.yaml:
```diff
@@ -156,28 +201,47 @@ classes:
+      free_text:
+        required: true
+      note_type_name:
+        required_true
```
required: true

markquintontulloch commented:
@chris-grove Good spots!  I've fixed the above comments, updated the ingest sets with a couple of missing allele associations, and also removed sequencing_status from the allele model.  Do you want me to change inheritance_mode to be a SlotAnnotation in this PR, or shall we do that as a separate PR?

chris-grove commented on model/schema/allele.yaml:
```diff
@@ -207,6 +356,13 @@ classes:
+        - allele_curie
+        - predicate_name
+        - cell_line_curie
+
```
because this is not inheriting from an 'evidence' slot-bearing class (like a hypothetical AssociationDTO class), 'evidence_curies' needs to be added here

chris-grove commented on model/schema/allele.yaml:
```diff
@@ -225,15 +381,13 @@ classes:
+      - allele_curie
+      - predicate_name
+      - image_curie
+      - primary_image
```
because this is not inheriting from an 'evidence' slot-bearing class (like a hypothetical AssociationDTO class), 'evidence_curies' needs to be added here

chris-grove commented on model/schema/allele.yaml:
```diff
@@ -245,18 +399,12 @@ classes:
-
+      - allele_curie
+      - predicate_name
+      - agm_curie
```
because this is not inheriting from an 'evidence' slot-bearing class (like a hypothetical AssociationDTO class), 'evidence_curies' needs to be added here

markquintontulloch commented:
@chris-grove - evidence_curies slots added.  We probably want to revise the Association and AssociationDTO classes at some point anyway to ensure the appropriate ranges are being used for predicates and to give them more descriptive names in the DTO classes.  Also AlleleGeneAssociation is currently inheriting an ROTerm predicate, which I think we decided yesterday was incorrect.

chris-grove commented:
@markquintontulloch OK sounds good! Yes, the ROTerm needs to be changed as the inherited predicate range (probably to VocabularyTerm) and the inheritance_mode needs to be a SlotAnnotation, but I/we can deal with that in our next iteration.

chris-grove commented on model/schema/allele.yaml:
```diff
@@ -225,15 +382,14 @@ classes:
+      - allele_curie
+      - predicate_name
+      - image_curie
+      - primary_image
```
'primary_image' slot range is not specified; seems that it should be a boolean

chris-grove commented on model/schema/allele.yaml:
```diff
@@ -324,12 +482,25 @@ slots:
     description: >-
```
'primary_image' slot range is not specified; seems that it should be a boolean

markquintontulloch replied:
Boolean seems sensible, but it’s just carried over from the current repo - not a new slot - so I’m not sure.  There’s definitely tidying up to do on the allele model but the aim of this PR was simply to transition to a model with the new paradigm of separate ingest classes.

markquintontulloch replied:
Are you happy enough with the new parts of the model that we can merge @chris-grove ?

chris-grove reviewed (approved):
Whew! That's quite the PR! I've gone through everything now and I think everything should be good.
